### PR TITLE
Update depth head

### PR DESCRIPTION
**Pull request by melisandeteng**
Use images from MiDaS model output 
predict inverse depth
add gradient matching loss term

### CODE REVIEW

File: omnigan/losses.py [hunk 1]
```diff
@@ -276,6 +276,23 @@ def __call__(self, prediction, target):
         relDiff = torch.mean(d) * torch.mean(d)
         return diff - relDiff
 
+class SIGMLoss(nn.Module):
+    def __init__(self, gmweight):
```
Comment (vict0rsch) [inline omnigan/losses.py:280]: add `device` argument and send `self.filter` to that device in the init, not to do it at every call

File: omnigan/losses.py [hunk 2]
```diff
@@ -276,6 +276,23 @@ def __call__(self, prediction, target):
...
+        super(SIGMLoss, self).__init__()
```
Comment (vict0rsch) [inline omnigan/losses.py:281]: it's identical but an easier way is just  `super().__init__()`

File: omnigan/losses.py [hunk 3]
```diff
@@ -276,6 +276,23 @@ def __call__(self, prediction, target):
...
+        sobely = torch.Tensor([[1,2,1],[0,0,0],[-1,-2,-1]])
+        self.filter = (0.5 * (sobelx + sobely))
+        
+    def __call__(self, prediction, target):
+        # get gradient map with sobel filters
+        batch_size = prediction.size()[0]
+        self.filter = (self.filter).expand((batch_size, 1, -1, -1)).to(device='cuda')      
+        grad_pred = F.conv2d(prediction, self.filter, stride=1)
+        grad_target = F.conv2d(target, self.filter, stride=1)
+        GMLoss = self.gmweight * torch.norm(grad_pred - grad_target)
```
Comment (vict0rsch) [inline omnigan/losses.py:293]: If it's not a class call, this should start lowercase. so perhaps `gm_loss`

File: omnigan/losses.py [hunk 4]
```diff
@@ -276,6 +276,23 @@ def __call__(self, prediction, target):
...
+        diff = SIMSELoss()(prediction, target) + GMLoss
```
Comment (vict0rsch) [inline omnigan/losses.py:294]: if `SIMSELoss` is initialized at every call, you might as well move this to the init (`self.simse_loss = SIMSELoss()`) and then in the call: `diff = self.simse_loss(prediction, target) + gm_loss`

File: omnigan/losses.py [hunk 5]
```diff
@@ -379,7 +396,7 @@ def get_losses(opts, verbose, device=None):
     # ? * add discriminator and gan loss to these task when no ground truth
     # ?   instead of noisy label
     if "d" in opts.tasks:
-        losses["G"]["tasks"]["d"] = SIMSELoss()
+        losses["G"]["tasks"]["d"] = SIGMLoss(opts.train.lambdas.G.d)
```
Comment (vict0rsch) [inline omnigan/losses.py:399]: I think defaults.yaml should look like: 

```yaml
train:
  lambdas: # scaling factors in the total loss
    G:
      d:
        main: 1 # the weight of the overall depth head to the generator's training
        gml: 1 # the weight of gml in the simse loss
```

**Review by vict0rsch — CHANGES_REQUESTED** [review]

File: omnigan/data.py [hunk 6]
```diff
@@ -242,8 +242,11 @@ def tensor_loader(path, task, domain):
         if Path(path).suffix == ".npy":
             arr = np.load(path)
         else:
-            arr = imread(path)  # .astype(np.uint8)
-        arr = torch.from_numpy(arr)
+            if domain == "r":
+                arr = np.array(Image.open(path).convert("L"))
```
Comment (alexrey88) [inline omnigan/data.py:246]: Did you check if the array made sense? When using MiDaS for smog, I tried convert("L") but it gave me images all white except a few black pixels only. What worked for me was:
`depth = Image.open(depth_path).convert("I")
   depth = np.array(depth)
    depth = normalize(depth)`

Comment (melisandeteng) [reply omnigan/data.py]: I actually did not need to convert as I just normlize the input

File: omnigan/losses.py (hunk 5 above)
Comment (melisandeteng) [reply omnigan/losses.py]: changed it!

File: omnigan/losses.py (hunk 1 above)
Comment (melisandeteng) [reply omnigan/losses.py]: done
